### PR TITLE
Split the instances by architecture.

### PR DESCRIPTION
**Pull request by illsilin**
CK used to build all instances for all architectures. Many of the original instances are generic and that's fine.
However, we now have several groups of instances which are architecture specific, such as xdl and wmma. 
These changes will make sure that the xdl instances get built only if the gfx9 architectures are on the target list and wmma instances will only get built if gfx11 are on the target list.

### CODE REVIEW

File: library/src/tensor_operation_instance/gpu/CMakeLists.txt [hunk 1]
```diff
@@ -124,6 +139,26 @@ FOREACH(subdir_path ${dir_list})
...
+        if(("${cmake_instance}" MATCHES "ONLY XDL_AND_DL_KERNELS") AND (NOT DEFINED DL_KERNELS) AND (NOT GPU_TARGETS MATCHES "gfx9"))
+            message("Found only xdl and dl instances, but gfx9 is not on the targets listand DL_KERNELS is not set. Skipping.")
+            set(add_inst 0)
+        endif()
+        if(("${cmake_instance}" MATCHES "ONLY XDL_AND_WMMA_KERNELS") AND (NOT GPU_TARGETS MATCHES "gfx11") AND (NOT GPU_TARGETS MATCHES "gfx9"))
+            message("Found only xdl and wmma instances, but gfx11 and gfx9 are not on the targets list. Skipping.")
+            set(add_inst 0)
+        endif()
+        if(("${cmake_instance}" MATCHES "XDL_DL_WMMA_KERNELS") AND (NOT GPU_TARGETS MATCHES "gfx11") AND (NOT GPU_TARGETS MATCHES "gfx9") AND (NOT DEFINED DL_KERNELS))
+            message("Found xdl, dl, and wmma instances, but none of those meet the target list. Skipping.")
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/CMakeLists.txt:159]: Do we actually need this with the changes you did to the `add_instance_library` function above?

**Review by aosewski — COMMENTED** [review]
In general the instance file filtering mechanism looks good. Maybe just the necessity of using `if result ... ` condition in CMake files is a bit inconvenient and counterintuitive.

**Review by carlushuang — APPROVED** [review]
LGTM

File: library/src/tensor_operation_instance/gpu/CMakeLists.txt (hunk 1 above)
Comment (illsilin) [reply library/src/tensor_operation_instance/gpu/CMakeLists.txt:159]: The add_instance_library contains the filters that determine whether or not to build an instance. While this piece of code goes over the list of the instance folders and decides which library to pack the instances to. If an instance is not built, we'll end up with linking errors in the end. That's why we have to filter the folders here.

**Comment by illsilin** [comment]
> In general the instance file filtering mechanism looks good. Maybe just the necessity of using `if result ... ` condition in CMake files is a bit inconvenient and counterintuitive.

Yes, that use of "result" variable is unfortunately the old style of cmake that we had to resort to because most of our official dockers are using cmake version 3.16. I was originally using the functions returning a value, but that functionality is only supported  in cmake 3.25+.

**Pull request merged**